divi/src/init.cpp

```diff
@@ -114,12 +114,6 @@ constexpr int nWalletBackups = 20;
  */
 #endif
 
-
-CChain chainActive;
-BlockMap mapBlockIndex;
-CBlockTreeDB* pblocktree = nullptr;
-CCoinsViewCache* pcoinsTip = nullptr;
-
```

Comment:
Given that the destruction of these objects had previously occurred at the end of a program during GC, it would seem possible that some thread out there somewhere may one day try to access these objects through the chain state manager after destruction via a race condition - low probability but not impossible. Maybe some safeguards to calling ChainstateManager::Get() would be required? Perhaps returning a shared_ptr instead? Thoughts?

Reply:
I think it is actually a feature to have (now) a well-defined life cycle.  A previous change already made it such that an access like this would `assert` fail, and all tests run fine - so hopefully there is no such situation.  But if it were, I think it would be much preferable to properly fix it (respecting the life cycle of `ChainstateManager`) instead of trying to work around it by using a shared pointer.

One of my next steps will be to actually remove the singleton "global" as well, and instead pass the one instance from `init` explicitly to every place it needs to be in.  That will then hopefully also make it more explicit and clear that no code accesses the `ChainstateManager` outside of its life cycle.

Reply:
All good ideas.

However, what I'm thinking is the following sequence of events:
1. ChainstateManager instance is created (Thread 0)
2. (Some time later...) ChainstateManager reference is accessed (Thread 1)
3. The ChainstateManager "unique_ptr" in init is deallocated due to destruction (Thread 2)
4. Dangling reference in (2) tries to access some component of ChainstateManger (Thread 1)

The mutex lock only guards against 'ordered accesses', but not all race conditions. In order to safely guard against race conditions, some manner of reference counting is needed.